subsys: usb: host: Add USB Host Vendor specific serial class support

Girinandha-M opened the pull request:
Dependency:

* https://github.com/zephyrproject-rtos/zephyr/pull/106802

Upstream Reference:

* https://github.com/zephyrproject-rtos/zephyr/pull/94504

subsys: usb: host: add vendor-specific serial class

Add USB vendor-specific serial host class driver that matches
devices by VID, PID.

This is not a standard CDC-ACM driver - it's for vendor-specific
serial devices that use bulk endpoints instead of CDC protocol.

Signed-off-by: Girinandha Manivelpandiyan <girinandha@linumiz.com>

Girinandha-M committed:
usb: host: add USB Serial support

add support for the vendor specific serial support class
to the usb subsys. This implementation enable the usb host class support
the vendor specific devices for serial communication.

Signed-off-by: Girinandha Manivelpandiyan <girinandha@linumiz.com>

josuah commented:
Hello @Girinandha-M, the fixes from https://github.com/zephyrproject-rtos/zephyr/pull/99334 got integrated into https://github.com/zephyrproject-rtos/zephyr/pull/94590.

If you wished to simplify your PR you may be able to switch to https://github.com/zephyrproject-rtos/zephyr/pull/94590.

Girinandha-M commented on subsys/usb/host/class/usbh_serial.c:
Thanks for the feedback — agreed on reusing the existing Zephyr serial/UART APIs.

The async UART API was chosen specifically to match modem-style usage (non-blocking TX/RX)

Please let me know if there are concerns with using the async UART API

carlescufi commented:
@Girinandha-M the dependencies should be merged now. Can you take this out of draft and rebase on top of `main`?

josuah commented on subsys/usb/host/class/Kconfig.serial:
```diff
@@ -0,0 +1,27 @@
+#
+
+config USBH_SERIAL
+	bool "USB Host Vendor-Specific Serial Support"
```
From discussions on Discord, this is the counterpart of this Linux driver:

https://github.com/torvalds/linux/blob/72c395024dac5e215136cbff793455f065603b06/drivers/usb/serial/usb-serial.c#L45-L50

With various vendor-specific extras, some device not having any vendor specifics listed there:

https://github.com/torvalds/linux/blob/72c395024dac5e215136cbff793455f065603b06/drivers/usb/serial/usb-serial-simple.c#L35-L127

Girinandha-M committed:
usb: host: assign device address before full descriptor read

Move alloc_device_address() and usbh_device_set_address()
before reading the full device descriptor.

the current sequence reads the full 18-byte device descriptor
at address 0 before assigning a device address. this causes
enumeration failure on the Quectel EG916Q-GL, which returns
bNumConfigurations=0 on the second GET_DESCRIPTOR request at
address 0.

reorder the sequence to assign the address after the initial
8-byte read, and fetch the full 18-byte descriptor using the
newly assigned address. the new order works around devices that
fail the second GET_DESCRIPTOR at address 0, while remaining
functional on devices that enumerate correctly with the
original order (verified on ECM dongle).

Signed-off-by: Girinandha Manivelpandiyan <girinandha@linumiz.com>

Girinandha-M commented:
Rebased on main, reworked based on feedback:

- Dropped DT bindings, using DEVICE_DEFINE + LISTIFY (same pattern as UVC)
- No public header, everything internal
- UART async API for modem/networking subsystem compatibility
- VID/PID filter matching with extensible quirk table
Tested on imxrt1064_evk + Quectel EG916Q: enumeration, class probe, AT command TX all working

Found and fixed a upstream HS enumeration bugs in usbh_device.c during testing (separate commit).

josuah reviewed (commented):
Here is some very incomplete review, in the meantime that a more complete review can be done.

Girinandha-M committed:
usb: host: fix interface skipping in descriptor walking

usbh_desc_get_next_function() sets skip_num=1 when starting from
a plain interface descriptor. usbh_desc_get_next() already advances
past the input, so an additional skip_num=1 makes the loop consume
one extra interface beyond the current one.

Devices with consecutive plain interfaces
(e.g. Quectel EG916Q-GL with 4 vendor interfaces) only get
ifaces 0 and 2 probed; 1 and 3 are silently skipped.

Drop the assignment. The IAD branch is unaffected since its skip_num
comes from bInterfaceCount, which counts interfaces grouped inside
the IAD, not the IAD descriptor itself.

Signed-off-by: Girinandha Manivelpandiyan <girinandha@linumiz.com>

tmon-nordic commented on subsys/usb/host/usbh_device.c:
```diff
@@ -561,31 +561,31 @@ int usbh_device_init(struct usb_device *const udev)
 	}
 
-	err = usbh_req_desc_dev(udev, sizeof(udev->dev_desc), &udev->dev_desc);
+	err = alloc_device_address(udev, &new_addr);
```
While this is OK from the interoperability point of view (allowing to work with broken devices), what exactly does "matching USB 2.0 enumeration sequence" refer to?

tmon-nordic reviewed (commented):
~~I think the first three commits are worthwhile to have in main even before there is agreement on how to proceed with "usb: host: add USB Serial support". Can you move them to separate PR?~~ Ah you already posted them separately in #106802

danielepa-telit commented:
Hey, thanks for this PR: it's very interesting and it would also be useful for supporting Telit modems and, in general, Qualcomm based USB modems.

Hopefully, in the near future I will give it a try.

josuah commented:
Removing "stale" label as it seems to just wait review

Girinandha-M commented:
@jfischer-no @josuah @tmon-nordic  Ping for Review. Thanks
